zipline/algorithm.py

```diff
@@ -457,30 +433,40 @@ def _create_generator(self, sim_params, source_filter=None):
         processed by the zipline, and False for those that should be
         skipped.
         """
-
-        if not self.initialized:
-            self.initialize(*self.initialize_args, **self.initialize_kwargs)
-            self.initialized = True
+        if sim_params is not None:
+            self.sim_params = sim_params
 
         if self.perf_tracker is None:
+            # Build a perf_tracker
+            self.perf_tracker = PerformanceTracker(
+                sim_params=self.sim_params,
+                env=self.trading_environment,
+                data_portal=self.data_portal)
+            # Set the dt initially to the period start by forcing it to change
+            self.on_dt_changed(self.sim_params.period_start)
+
             # HACK: When running with the `run` method, we set perf_tracker to
             # None so that it will be overwritten here.
             self.perf_tracker = PerformanceTracker(
```

Comment:
Why do we set the perf tracker twice?

What are the cases that the `if self.perf_tracker is None` condition is testing? Looks like `__init__` sets it to `None` always?  If we call `run` multiple times, do we want to re-use the perf tracker?